discovery/uyuni/uyuni.go

```diff
@@ -0,0 +1,343 @@
+		managedGroupNames = append(managedGroupNames, systemGroupInfo.GroupName)
+	}
+
+	if len(managedGroupNames) == 0 {
+		managedGroupNames = []string{"No group"}
+	}
+	return managedGroupNames
+}
+
+func (d *Discovery) getTargetsForSystems(
+	rpcClient *xmlrpc.Client,
+	token string,
+	systemGroupIDsBySystemID map[int][]systemGroupID,
+) ([]model.LabelSet, error) {
+
+	result := make([]model.LabelSet, 0)
+
+	systemIDs := make([]int, 0, len(systemGroupIDsBySystemID))
+	for systemID := range systemGroupIDsBySystemID {
+		systemIDs = append(systemIDs, systemID)
+	}
+
+	endpointInfos, err := getEndpointInfoForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get endpoints information")
+	}
+
+	networkInfoBySystemID, err := getNetworkInformationForSystems(rpcClient, token, systemIDs)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get the systems network information")
+	}
+
+	for _, endpoint := range endpointInfos {
+		systemID := endpoint.SystemID
+		labels := d.getEndpointLabels(
+			endpoint,
+			systemGroupIDsBySystemID[systemID],
+			networkInfoBySystemID[systemID])
+		result = append(result, labels)
+
+		if networkInfoBySystemID[systemID].Hostname != "" {
+			level.Debug(d.logger).Log("msg", "Found endpoint",
+				"Host", networkInfoBySystemID[systemID].Hostname,
+				"PrimaryFQDN", networkInfoBySystemID[systemID].PrimaryFQDN,
+				"Network", fmt.Sprintf("%+v", networkInfoBySystemID[systemID]),
+				"Groups", fmt.Sprintf("%+v", systemGroupIDsBySystemID[systemID]))
+		}
+	}
+
+	return result, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	cfg := d.sdConfig
+	apiURL := cfg.Host + uyuniXMLRPCAPIPath
+
+	startTime := time.Now()
+
+	// Check if the URL is valid and create rpc client
+	_, err := url.ParseRequestURI(apiURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "Uyuni Server URL is not valid")
+	}
+
+	rpcClient, err := xmlrpc.NewClient(apiURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer rpcClient.Close()
+
+	token, err := login(rpcClient, cfg.User, string(cfg.Pass))
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to login to Uyuni API")
+	}
+	defer func() {
+		if err := logout(rpcClient, token); err != nil {
+			level.Warn(d.logger).Log("msg", "Failed to log out from Uyuni API", "err", err)
+		}
+	}()
+
+	systemGroupIDsBySystemID, err := getSystemGroupsInfoOfMonitoredClients(rpcClient, token)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get the managed system groups information of monitored clients")
+	}
+
+	targets := make([]model.LabelSet, 0)
+	if len(systemGroupIDsBySystemID) > 0 {
+		targetsForSystems, err := d.getTargetsForSystems(rpcClient, token, systemGroupIDsBySystemID)
+		if err != nil {
+			return nil, err
+		}
+		targets = append(targets, targetsForSystems...)
+		level.Info(d.logger).Log("msg", "Total discovery time", "time", time.Since(startTime))
```

Comment:
Why do we need this? This is also very verbose.

Reply:
Removed.